### ISSUE TITLE
Why the training performance can't be replicated

### ISSUE THREAD

**Issue by Chen94yue** [KevinMusgrave/pytorch-metric-learning#99]
My code for train:
```python
def visualizer_hook(umapper, umap_embeddings, labels, split_name, keyname):
    logging.info("UMAP plot for the {} split and label set {}".format(split_name, keyname))
    label_set = np.unique(labels)
    num_classes = len(label_set)
    fig = plt.figure(figsize=(20,15))
    plt.gca().set_prop_cycle(cycler("color", [plt.cm.nipy_spectral(i) for i in np.linspace(0, 0.9, num_classes)]))
    for i in range(num_classes):
        idx = labels == label_set[i]
        plt.plot(umap_embeddings[idx, 0], umap_embeddings[idx, 1], ".", markersize=1)
    plt.savefig(cfg.LOGGER.VISDIR)



def main(cfg):
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    trunk = torchvision.models.resnet50(pretrained=True)
    trunk.requires_grad_(False)
    trunk = torch.nn.DataParallel(trunk.to(device))
    embedder = torch.nn.DataParallel(Linear([1000, 256]).to(device))

    trunk_optimizer = torch.optim.Adam(trunk.parameters(), lr=cfg.SOLVER.BASE_LR, weight_decay=cfg.SOLVER.WEIGHT_DECAY)
    embedder_optimizer = torch.optim.Adam(embedder.parameters(), lr = cfg.SOLVER.BASE_LR * 10, weight_decay=cfg.SOLVER.WEIGHT_DECAY)
    trunk_lr_scheduler = lr_scheduler.StepLR(trunk_optimizer, step_size=15, gamma=0.1)
    embedder_lr_scheduler = lr_scheduler.StepLR(embedder_optimizer, step_size=15, gamma=0.1)
    train_transform = set_transforms(cfg, True)
    val_transform = set_transforms(cfg, False)

    train_data = imagenet_data(cfg, train_transform, True)
    print("Load train done.")
    val_data = imagenet_data(cfg, val_transform, False)
    print("Load val done.")
    loss = losses.MultiSimilarityLoss(cfg.LOSSES.MULTI_SIMILARITY_LOSS.SCALE_POS, cfg.LOSSES.MULTI_SIMILARITY_LOSS.SCALE_NEG)
    # loss = losses.TripletMarginLoss(margin=0.1)
    miner = miners.MultiSimilarityMiner(epsilon=cfg.LOSSES.MULTI_SIMILARITY_LOSS.HARD_MINING)
    sampler = samplers.MPerClassSampler(train_data.labels, m=cfg.DATA.NUM_PRE_INSTANCES, length_before_new_iter=len(train_data))

    models = {'trunk': trunk, "embedder": embedder}
    optimizers = {"trunk_optimizer": trunk_optimizer,"embedder_optimizer": embedder_optimizer}
    loss_funcs = {"metric_loss": loss}
    mining_funcs = {"tuple_miner": miner}
    lr_schedulers = {"trunk_scheduler_by_epoch": trunk_lr_scheduler, "embedder_scheduler_by_epoch": embedder_lr_scheduler}
    record_keeper, _, _ = logging_presets.get_record_keeper(cfg.LOGGER.LOGDIR, cfg.LOGGER.THESORBOARDDIR)
    hooks = logging_presets.get_hook_container(record_keeper)
    dataset_dict = {"val": val_data}
    model_folder = cfg.SAVE_DIR

    tester = testers.GlobalEmbeddingSpaceTester(end_of_testing_hook = hooks.end_of_testing_hook,
                                                 visualizer = umap.UMAP(),
                                                 visualizer_hook = visualizer_hook,
                                                 # use_trunk_output=True,
                                                 dataloader_num_workers = cfg.DATA.NUM_WORKERS)


    end_of_epoch_hook = hooks.end_of_epoch_hook(tester,
                                                dataset_dict,
                                                model_folder,
                                                )

    print("Init trainer.")
    trainer = trainers.MetricLossOnly(models,
                                      optimizers,
                                      cfg.DATA.TRAIN_BATCHSIZE,
                                      loss_funcs,
                                      mining_funcs,
                                      train_data,
                                      lr_schedulers = lr_schedulers,
                                      sampler = sampler,
                                      freeze_trunk_batchnorm = True,
                                      dataloader_num_workers = cfg.DATA.NUM_WORKERS,
                                      end_of_iteration_hook = hooks.end_of_iteration_hook,
                                      end_of_epoch_hook = end_of_epoch_hook)
    print("Init Done!")
    trainer.train(num_epochs=cfg.SOLVER.EPOCH)

def parse_args():
    """
    Parse input arguments
    """
    parser = argparse.ArgumentParser(description='Train a retrieval network')
    parser.add_argument(
        '--cfg',
        dest='cfg_file',
        help='config file',
        default=None,
        type=str
        )
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    main(cfg)
```
My code for test:
```python
def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    trunk = torchvision.models.resnet50(pretrained=True)
    trunk = torch.nn.DataParallel(trunk.to(device))
    embedder = Linear([1000, 256])
    embedder_pth = 'saved_models/embedder_best.pth'
    embedder.load_state_dict(torch.load(embedder_pth, map_location=device))
    embedder = torch.nn.DataParallel(embedder.to(device))

    val_transform = set_transforms(cfg, False)
    val_data = imagenet_data(cfg, val_transform, False)

    data_loader = torch.utils.data.DataLoader(val_data, batch_size=320, shuffle=False, num_workers=12)

    with torch.no_grad():
        Ff = torch.zeros(len(val_data), 256).cuda()
        index = 0
        label_list = []
        for images, labels in tqdm(data_loader):

            label_list += [i.item() for i in labels]
            x = images.cuda()
            features_pretrain = trunk(x)
            features_finetune = embedder(features_pretrain)
            features_finetune = F.normalize(features_finetune, p=2, dim=1)
            Ff[index:index + len(labels)] = features_finetune
            index += len(labels)
    acc = AccuracyCalculator(include=(), exclude=(), average_per_class=False)
    print(acc.get_accuracy(query = Ff.cpu().numpy(), reference = Ff.cpu().numpy(), query_labels =  np.array(label_list), reference_labels = np.array(label_list), embeddings_come_from_same_source=True))
```

the result in training los is:
epoch,AMI_level0,NMI_level0,mean_average_precision_at_r_level0,precision_at_1_level0,r_precision_level0,best_epoch,best_accuracy
23,0.6148627020563618,0.7620803255750865,0.2960472842785475,0.63168,0.38713714285714285,23,0.2960472842785475

but the result in testing:
{'AMI': 0.5835943752042939, 'NMI': 0.7389371528303541, 'mean_average_precision_at_r': 0.26242998784948846, 'precision_at_1': 0.60576, 'r_precision': 0.3523061224489796}

**Comment by KevinMusgrave** [comment]
In your main() testing function, can you set your models to eval() mode and see if that changes anything?

**Comment by Chen94yue** [comment]
Sorry, I forgot this detial.
